### PR TITLE
Fix serialization of FileTarget

### LINKED ISSUE #81: [BUG] dotter init fails when any files are present in CWD
## Environment
- **OS**: Arch Linux up-to-date as of 2021-08-09
- **Dotter version**: Dotter 0.12.5

## Description
When `dotter init` is run in a directory containing any file besides a `.dotter` directory, the command fails with the following output:
```console
[ERROR] Failed to initalize directory
Caused by:
    save dummy config
    save global config
    serialize data
    unsupported Rust type
```

## Reproduction
In a suitable temporary directory, run:
```sh
touch file
dotter init
```

## Additional notes
I have tracked down the source of this bug and have submitted a pull request (#82).

I'm not sure how this was missed until now; maybe it was a regression and `dotter init` hasn't been tested since it was first implemented.

### PR DESCRIPTION
**Pull request by Diomendius**
Fixes #81

In investigating #81, I found that `config::FileTarget` has a custom `Deserialize` impl because it doesn't cleanly map the desired TOML structure to any of Serde's configurable representations; `FileTarget::Automatic("foo".into())` should map to `"foo"` (untagged representation) but an internally tagged representation should be used for the other variants: `{ type = "symbolic", target = "foo" }`

Serde does not support mixed enum representation like this, so the closest you can get is `{ type = "automatic", target = "foo" }`, hence a custom impl was written. For some reason, no `Serialize` impl was written (probably because AFAIK only `dotter init` uses it currently, but this means that the serialized representation doesn't match what the program expects to deserialize.

This is a bug on its own, but because the `toml` crate is used, Dotter never actually reaches this state because `toml` fails at runtime when trying to serialize a `FileTarget`, as the default representation is "externally tagged", which `toml` does not yet support. The TOML format does support this representation (which could render as `{ "symbolic" = { target = "foo" } }` and as part of a `Files` entry would look like this monstrosity: `"source" = { "symbolic" = { target = "foo" } }`) but regardless, it's not very ergonomic, nor is it intended.

Because writing a corresponding custom `Serialize` impl seemed like far too much boilerplate, I came up with what I would judge as a better solution. I lifted the proxy `FileTarget` representation (`This`) out of the custom `Deserialize` impl, deleted said impl and split this representation into two enums, `FileTargetOuterRepr` and `FileTargetInnerRepr` (feel free to choose better names if you have any ideas), the outer enum annotated with `#[serde(untagged)` and the inner with `#[serde(tag = "type")]`.

I then annotated `FileTarget` itself with `#[serde(from = "FileTargetOuterRepr", into = "FileTargetOuterRepr")]` so Serde would convert to/from the proxy type as needed, and added the corresponding `From` impls. I would have preferred even less boilerplate than this, but it's much more elegant than two mostly identical copies of the original `Deserialize` impl. Maybe someday Serde will support mixed representations where some (unambiguous) variants are left untagged.

## Other notes

Tracking this down was certainly not without friction. The `anyhow` error Dotter spat out informed me the error originated from somewhere in `toml::to_string`, but it quickly became evident the useful information would have been somewhere in Serde's generated `Serialize` impl and even viewing expanded macros is difficult, let alone getting a debugger to show said expansions. I've since looked for some improvements to my NeoVim configuration and should now be able to use rust-analyzer's macro expansion features, which should alleviate the headaches somewhat.

It should be possible to compile the expanded macros after saving them to a scratch working tree which should let a debugger show those lines. As usual, hindsight is 20/20. Rust's macros are still basically debugger-impenetrable using only the standard Cargo + rustc tools, though. No, `rustc -X trace_macros=y` doesn't count, `rustc -X trace_macros=y` is basically the magnetized-needle-and-a-steady-hand of Rust macro debugging tools.

I think the problem I described in the commit message (8cdf0b7457f91dc4553c7fb22b8ddcbfccd83ba8) relates more to a problem I had when trying different solutions to this problem in a test program than to Dotter as originally written. Dotter's code as written runs into a limitation of `toml`, but if `#[serde(tag = "type")]` is added to `FileTarget` Serde will error at runtime instead, saying "cannot serialize tagged newtype variant".

This refers to `FileTarget::Automatic` being a newtype variant wrapping a `PathBuf`. I think this refers to internal tagging specifically, but isn't worded clearly. I think it fails because the representation would be `{ type = "automatic", "foo/bar" }`, where the contained type represents a value with no key (or a key with no value).

## TL;DR
`FileTarget` had a bad `Serialize` impl which compiled but crashed at runtime. I fixed this and refactored some code in the process.

I then wrote a lot about how I did it and how confused I got at the time, and confused myself more in the process of writing about it.

And all for the want of a `#[serde(untagged_variant)]`.

### CODE REVIEW

**Comment by SuperCuber** [comment]
Very clever implementation, and yes you guessed correctly (there was an even longer Deserialize implementation earlier and the current implementation was contributed).

I never tried digging into macros. It's hard enough to figure out why stuff doesn't compile sometimes with heavy macros/generics, I can't imagine trying to figure out why it doesn't run :P

Worry not about clippy CI failing by the way
